### PR TITLE
[Aptos Framework] [Account] Added signer capability delegation functions

### PR DESCRIPTION
**Pull request by chloeqjz**
### Description
Added signer capability functions that allow users to offer signer capability to another account, and accept signer capability from another account.

### Test Plan
Unit tests.

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3293)

### CODE REVIEW

**Comment by areshand** [comment]
can we remove the `store` from signer capability?

we can provide function call "acquire_signer_capability(account: &signer): SignerCapability" in account.move to return the signer cap when account with offered signer capability  want to use the signer cap.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 1]
```diff
@@ -329,6 +334,52 @@ module aptos_framework::account {
         rotation_capability
     }
 
+    // Offer the account owner's signer capablity to another account
+    public entry fun offer_signer_capability_ed25519(account: &signer, signer_capability_sig_bytes: vector<u8>, account_public_key_bytes: vector<u8>, recipient_address: address) acquires Account {
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/account.move:338]: why do we want to use sig bytes?

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 2]
```diff
@@ -329,6 +334,52 @@ module aptos_framework::account {
...
+        // Verify a digital-signature-based capability that assures us this signer capability offer was intended by the account owner
+        assert!(ed25519::signature_verify_strict_t(&signer_capability_sig, &pubkey, signer_capability_offer_proof_challenge), EINVALID_PROOF_OF_KNOWLEDGE);
+
+        // Add the recipient's address in account owner's signer capability offer once we verify that this action is intended by the account owner
+        option::fill(&mut account_resource.signer_capability_offer.for, recipient_address);
+    }
+
+
+    // Accept signer capability from `offerer_address` if there's an existing signer capability offer to the account owner
+    public fun accept_signer_capability_ed25519(account: &signer, offerer_address: address) : SignerCapability acquires Account {
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/account.move:364]: once we remove store from signer_cap, we don't need this function.

Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:364]: the problem with removing store from `signer_cap` is that we're storing the signer capability in ResourceAccount: https://github.com/aptos-labs/aptos-core/blob/d345a2007c95443b1eabf673257d508b439ab399/aptos-move/framework/aptos-framework/sources/resource_account.move#L48

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 3]
```diff
@@ -329,6 +334,52 @@ module aptos_framework::account {
...
+        assert!(verify_authentication_key_matches_ed25519_public_key(account_resource.authentication_key, account_public_key_bytes), EWRONG_CURRENT_PUBLIC_KEY);
+
+        //  Construct a SignerCapabilityOfferProofChallenge struct
+        let signer_capability_offer_proof_challenge = SignerCapabilityOfferProofChallenge {
+            sequence_number: account_resource.sequence_number,
+            recipient_address,
+        };
+
+        // Verify a digital-signature-based capability that assures us this signer capability offer was intended by the account owner
+        assert!(ed25519::signature_verify_strict_t(&signer_capability_sig, &pubkey, signer_capability_offer_proof_challenge), EINVALID_PROOF_OF_KNOWLEDGE);
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/account.move:356]: Wrap error

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 4]
```diff
@@ -329,6 +334,52 @@ module aptos_framework::account {
...
+    public entry fun offer_signer_capability_ed25519(account: &signer, signer_capability_sig_bytes: vector<u8>, account_public_key_bytes: vector<u8>, recipient_address: address) acquires Account {
+        let addr = signer::address_of(account);
+        assert!(exists_at(addr) && exists_at(recipient_address), error::not_found(EACCOUNT_DOES_NOT_EXIST));
+
+        let pubkey = ed25519::new_unvalidated_public_key_from_bytes(account_public_key_bytes);
+        let signer_capability_sig = ed25519::new_signature_from_bytes(signer_capability_sig_bytes);
+
+        // Get the current authentication key of the account and verify that it matches with `account_public_key_bytes`
+        let account_resource = borrow_global_mut<Account>(addr);
+        assert!(verify_authentication_key_matches_ed25519_public_key(account_resource.authentication_key, account_public_key_bytes), EWRONG_CURRENT_PUBLIC_KEY);
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/account.move:347]: Wrap error

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 5]
```diff
@@ -329,6 +334,52 @@ module aptos_framework::account {
...
+    // Offer the account owner's signer capablity to another account
+    public entry fun offer_signer_capability_ed25519(account: &signer, signer_capability_sig_bytes: vector<u8>, account_public_key_bytes: vector<u8>, recipient_address: address) acquires Account {
+        let addr = signer::address_of(account);
+        assert!(exists_at(addr) && exists_at(recipient_address), error::not_found(EACCOUNT_DOES_NOT_EXIST));
+
+        let pubkey = ed25519::new_unvalidated_public_key_from_bytes(account_public_key_bytes);
+        let signer_capability_sig = ed25519::new_signature_from_bytes(signer_capability_sig_bytes);
+
+        // Get the current authentication key of the account and verify that it matches with `account_public_key_bytes`
+        let account_resource = borrow_global_mut<Account>(addr);
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/account.move:346]: nit: Why declare these variables all the way up here and not where they're needed, closer to line 355?

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 2 above)
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/account.move:364]: So this is only meant to be called from a module or script? The signer cap needs to be stored somewhere and can't just be given directly to an account

**Review by davidiw — COMMENTED** [review]
I want to think about the accept side before we enable it. Let's maybe table this PR and focus first on the rotation.

I'm half minded to never allow these to be storeable and only as used in its current form. We could probably keep the offer api and then defer on the claim side... make using only possible, for example.

**Comment by davidiw** [comment]
So based upon Bo's proposal, is there some way that we could expose this that mirrors the goals of that proposal. Explicitly, we could imagine that the signer can never be exported from the Offer registry and that this acts more as an ACL.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 6]
```diff
@@ -329,6 +334,52 @@ module aptos_framework::account {
...
+        assert!(option::contains(&account_resource.signer_capability_offer.for, &addr), EINVALID_ACCEPT_CAPABILITY_OFFER);
+
+        // If there's an existing signer capability offer for this account in the offerer's account,
+        // we create a SignerCapability of offerer and return the SignerCapability
+        let signer_capability = SignerCapability {
+            account: offerer_address,
+        };
+        option::extract(&mut account_resource.signer_capability_offer.for);
+
+        signer_capability
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:379]: If we want to expose this, my preference would be that we only expose a signer and not the capability. That is we NEVER allow exporting of the SignerCapability out of this file -- for now. I'd be happy to approve after that. I think that also aligns with @areshand design goals.

Comment (areshand) [reply aptos-move/framework/aptos-framework/sources/account.move:379]: Yes, will work with @chloeqjz  on revising this. I think the idea is to not return signer capability and limit its usage to existing modules.

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 1 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move]: signature bytes is the signature by the account's private key on the SignerCapabilityOfferChallenge to indicates that the account owner intends to delegate their signer capability, so it's different from the signer

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 7]
```diff
@@ -342,6 +352,75 @@ module aptos_framework::account {
         rotation_capability
     }
 
+    public entry fun offer_signer_capability(account: &signer, signer_capability_sig_bytes: vector<u8>, account_scheme: u8, account_public_key_bytes: vector<u8>, recipient_address: address) acquires Account {
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/account.move:355]: 1. let's call it `offer_signer_authority`
2. signer_capability_sig_bytes rename to og_signer_proof_bytes?

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 8]
```diff
@@ -342,6 +352,75 @@ module aptos_framework::account {
...
+        assert!(exists_at(addr) && exists_at(recipient_address), error::not_found(EACCOUNT_DOES_NOT_EXIST));
+        assert!(account_scheme == 0 || account_scheme == 1, error::invalid_argument(EINVALID_SCHEME));
+
+        let account_resource = borrow_global_mut<Account>(addr);
+        let signer_capability_offer_proof_challenge = SignerCapabilityOfferProofChallenge {
+            sequence_number: account_resource.sequence_number,
+            recipient_address,
+        };
+
+        if (account_scheme == ED25519_SCHEME) {
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/account.move:366]: why do we need to distinguish multi-sig and single sig here?  My understanding this is signer authority delegation. the signer is the same for multi-sig and single sig. And the proof of og signer owner bytes would be the same as well.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 9]
```diff
@@ -342,6 +352,75 @@ module aptos_framework::account {
...
+            assert!(multi_ed25519::signature_verify_strict_t(&signer_capability_sig, &pubkey, signer_capability_offer_proof_challenge), error::invalid_argument(EINVALID_PROOF_OF_KNOWLEDGE));
+            if (option::is_some(&account_resource.signer_capability_offer.for)) {
+                option::swap(&mut account_resource.signer_capability_offer.for, recipient_address);
+            } else {
+                option::fill(&mut account_resource.signer_capability_offer.for, recipient_address);
+            };
+        }
+    }
+
+    public fun accept_signer_capability(account: &signer, offerer_address: address) acquires Account, SignerAuthority {
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/account.move:394]: rename it to` accept_signer_authority`

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 10]
```diff
@@ -342,6 +352,75 @@ module aptos_framework::account {
...
+        assert!(exists_at(offerer_address), error::not_found(EACCOUNT_DOES_NOT_EXIST));
+
+        let account_resource = borrow_global_mut<Account>(offerer_address);
+        let addr = signer::address_of(account);
+        assert!(option::contains(&account_resource.signer_capability_offer.for, &addr), error::invalid_argument(EINVALID_ACCEPT_CAPABILITY_OFFER));
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/account.move:399]: we don't need this anymore. acquire_signer should only need to use SignerAuthority

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 11]
```diff
@@ -65,6 +65,16 @@ module aptos_framework::account {
         recipient_address: address,
     }
 
+    struct SignerCapabilityOfferProofChallenge has drop {
+        sequence_number: u64,
+        recipient_address: address,
+    }
+
+    struct SignerAuthority has key {
+        offer: Table<address, CapabilityOffer<SignerCapability>>,
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/account.move:74]: this should be `offer: CapabilityOffer<SignerCapability>,`

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 8 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:389]: because we wanted to make sure that the public key matches the account's auth key, and the methods for creating `new_unvalidated_public_key_from_bytes` and `unvalidated_public_key_to_authentication_key` are different for ed25519 and multi-ed25519 key?

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 7 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:376]: think we are going to defer the `SignerAuthority` solution to the future / longer-term plans for now

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 12]
```diff
@@ -20,6 +20,8 @@ module aptos_framework::account {
     friend aptos_framework::resource_account;
     friend aptos_framework::transaction_validation;
 
+//    use std::debug;
+
```
Comment (gregnazario) [inline aptos-move/framework/aptos-framework/sources/account.move:24]: Please remove

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 7 above)
Comment (areshand) [reply aptos-move/framework/aptos-framework/sources/account.move:376]: could you help us understand what the plan is now? @davidiw  
I thought we still want to introduce signer authority and NOT to deprecate existing usage of signer_capability.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 13]
```diff
@@ -320,10 +329,10 @@ module aptos_framework::account {
         assert!(ed25519::signature_verify_strict_t(&rotation_capability_sig, &pubkey, rotation_capability_offer_proof_challenge), EINVALID_PROOF_OF_KNOWLEDGE);
 
         // Add the recipient's address in account owner's rotation capability offer once we verify that this action is intended by the account owner
-        option::fill(&mut account_resource.rotation_capability_offer.for, recipient_address);
+         update_or_add_capability_offer<RotationCapability>(addr, recipient_address);
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:332]: weird white space

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 14]
```diff
@@ -342,6 +351,73 @@ module aptos_framework::account {
...
+        } else if (type_info::type_of<T>() == type_info::type_of<RotationCapability>()) {
+            if (option::is_some(&account_resource.rotation_capability_offer.for)) {
+                option::swap(&mut account_resource.rotation_capability_offer.for, recipient_address);
+            } else {
+                option::fill(&mut account_resource.rotation_capability_offer.for, recipient_address);
+            };
+        } else {
+            abort error::invalid_argument(EINVALID_CAPABILITY_TYPE)
+        };
+    }
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:373]: This is actual more expensive and not saving much in terms of reusability. It would be better to inline this afaiu.

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 8 above)
Comment (areshand) [reply aptos-move/framework/aptos-framework/sources/account.move:389]: I see. my confusion is why we need this guard. Could you help me understand the reasoning?
Previously, key rotation requires this proof as we don't want signer delegators to acquire the entire account in one txn.

Now, suppose a malicious signer delegator want to provide his signer to another one. I don't see the risk is higher if we allow this.

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 7 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/account.move:376]: welcome to the hell of naming :)

@areshand is right in that you cannot retrieve this capability, so it remains an authority.
@chloeqjz is right in that this does create a capability, but we only allow it to be used as an authority.

Also we named everything else capability.

My gut feeling is that we should probably continue along with capability but note that it is currently only useful as an authority.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 15]
```diff
@@ -342,6 +351,73 @@ module aptos_framework::account {
...
+
+        // verify that the `SignerCapabilityOfferProofChallenge` is correct and signed by the account owner's private key
+        if (account_scheme == ED25519_SCHEME) {
+            let pubkey = ed25519::new_unvalidated_public_key_from_bytes(account_public_key_bytes);
+            let matching_auth_key_to_public_key_bytes = ed25519::unvalidated_public_key_to_authentication_key(&pubkey);
+            assert!(account_resource.authentication_key == matching_auth_key_to_public_key_bytes, error::invalid_argument(EWRONG_CURRENT_PUBLIC_KEY));
+
+            let signer_capability_sig = ed25519::new_signature_from_bytes(signer_capability_sig_bytes);
+            assert!(ed25519::signature_verify_strict_t(&signer_capability_sig, &pubkey, signer_capability_offer_proof_challenge), error::invalid_argument(EINVALID_PROOF_OF_KNOWLEDGE));
+        } else {
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:396]: let's not use an else, let's use an else if and an abort.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 16]
```diff
@@ -342,6 +351,73 @@ module aptos_framework::account {
...
+            let signer_capability_sig = multi_ed25519::new_signature_from_bytes(signer_capability_sig_bytes);
+            assert!(multi_ed25519::signature_verify_strict_t(&signer_capability_sig, &pubkey, signer_capability_offer_proof_challenge), error::invalid_argument(EINVALID_PROOF_OF_KNOWLEDGE));
+        };
+
+        // update the existing signer capability offer or put in a new signer capability offer for the current account
+        update_or_add_capability_offer<SignerCapability>(addr, recipient_address);
+    }
+
+    /// Accept signer capability from `offerer_address` if there's an existing signer capability offer to the account owner
+    public fun accept_signer_capability(account: &signer, offerer_address: address) : signer acquires Account {
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:410]: create_authorized_signer -- this is where @areshand's comments can be better leveraged.

**Review by davidiw — COMMENTED** [review]
logically correct, just naming and coding improvements....

great work!!!!!!

next up is to do the same with the rotation capability that we dropped...

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 14 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move]: ooh i see - just updated!

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 15 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:411]: I was thinking bc we have `assert!(account_scheme == ED25519_SCHEME || account_scheme == MULTI_ED25519_SCHEME, error::invalid_argument(EINVALID_SCHEME));` at the start of this function, it should be an `else`?

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 16 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move]: okie just updated! thanks so much for reviewing : )

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 17]
```diff
@@ -299,47 +306,137 @@ module aptos_framework::account {
         account_public_key_bytes: vector<u8>,
         recipient_address: address,
     ) acquires Account {
+        offer_rotation_capability(account, rotation_capability_sig_bytes, ED25519_SCHEME, account_public_key_bytes, recipient_address);
+    }
+
+    /// Accept rotation capability from `offerer_address` if there's an existing rotation capability offer to the account owner in the offerer's account
+    public fun accept_rotation_capability_ed25519(account: &signer, offerer_address: address) : RotationCapability acquires Account {
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:313]: Let's separate out the rotation functionality to another PR. As this creates other issues for us to consider.

Also there's nothing about ed25519 key here.

**Review by davidiw — COMMENTED** [review]
* Let's nail down the naming, since we cannot rename later
* Let's not accept rotation capability -- we probably want to do something like we do with signers -- leave it as an ACL -- but this PR should really just be about one thing, it is currently three.
* I'm okay with the updates to offer_rotaton_capability, but this really should be 2 PRs.
* Eliminate the early check on schemes

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 18]
```diff
@@ -299,47 +306,137 @@ module aptos_framework::account {
...
+        // update the existing signer capability offer or put in a new signer capability offer for the current account
+        if (option::is_some(&account_resource.signer_capability_offer.for)) {
+            option::swap(&mut account_resource.signer_capability_offer.for, recipient_address);
+        } else {
+            option::fill(&mut account_resource.signer_capability_offer.for, recipient_address);
+        };
+    }
+
+    /// Return a signer of the offerer, if there's an existing signer capability offer at the offerer's address
+    public fun check_signer_capability_and_create_authorized_signer(account: &signer, offerer_address: address) : signer acquires Account {
```
[the remark this answers is not in the document]
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:431]: oh i mean i liked the name, but thought we need to distinguish the case of rotation cap and the case of signer cap? or we can just pass in the type?

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 17 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:313]: will move all rotation cap-related changes to another pr - but I think this function already exists on the main branch, and removing it will cause back-ward incompatibility? not sure if people are using this tho

Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/account.move:334]: I guess we did, I guess we did... then I guess this is what it is...

let's contemplate if we want to put an abort here for now... but in another PR.

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 15 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/account.move:379]: add the else if / else.

**Review by davidiw — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 19]
```diff
@@ -342,6 +349,63 @@ module aptos_framework::account {
         rotation_capability
     }
 
+    /// Cffer signer capability to another address
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:352]: Typo: Cffer should be Offer

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 20]
```diff
@@ -342,6 +349,63 @@ module aptos_framework::account {
...
+        account_scheme: u8,
+        account_public_key_bytes: vector<u8>,
+        recipient_address: address
+    ) acquires Account {
+        let addr = signer::address_of(account);
+        assert!(exists_at(addr) && exists_at(recipient_address), error::not_found(EACCOUNT_DOES_NOT_EXIST));
+
+        let account_resource = borrow_global_mut<Account>(addr);
+        // proof that this account intends to delegate its signer capability to another account
+        let signer_capability_offer_proof_challenge = SignerCapabilityOfferProofChallenge {
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:365]: Given that there's only one proof challenge in this function, the long name is not necessary: I'd just call this `proof_chal[lenge]`

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 21]
```diff
@@ -342,6 +349,63 @@ module aptos_framework::account {
...
+        if (account_scheme == ED25519_SCHEME) {
+            let pubkey = ed25519::new_unvalidated_public_key_from_bytes(account_public_key_bytes);
+            let matching_auth_key_to_public_key_bytes = ed25519::unvalidated_public_key_to_authentication_key(&pubkey);
+            assert!(account_resource.authentication_key == matching_auth_key_to_public_key_bytes, error::invalid_argument(EWRONG_CURRENT_PUBLIC_KEY));
+
+            let signer_capability_sig = ed25519::new_signature_from_bytes(signer_capability_sig_bytes);
+            assert!(ed25519::signature_verify_strict_t(&signer_capability_sig, &pubkey, signer_capability_offer_proof_challenge), error::invalid_argument(EINVALID_PROOF_OF_KNOWLEDGE));
+        } else if (account_scheme == MULTI_ED25519_SCHEME) {
+            let pubkey = multi_ed25519::new_unvalidated_public_key_from_bytes(account_public_key_bytes);
+            let matching_auth_key_to_public_key_bytes = multi_ed25519::unvalidated_public_key_to_authentication_key(&pubkey);
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:380]: Same here/

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 19 above)
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:352]: This doesn't clarify who's offering to whom. A more instructive comment would be:

 > Offers the capability to sign on behalf of `account` to the account at address `recipient_address`.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 22]
```diff
@@ -342,6 +349,63 @@ module aptos_framework::account {
...
+    }
+
+    /// Return a signer of the offerer, if there's an existing signer/rotation capability offer at the offerer's address
+    public fun create_authorized_signer(account: &signer, offerer_address: address) : signer acquires Account {
+        assert!(exists_at(offerer_address), error::not_found(EACCOUNT_DOES_NOT_EXIST));
+
+        // Check if there's an existing signer capability offer from the offerer
+        let account_resource = borrow_global_mut<Account>(offerer_address);
+        let addr = signer::address_of(account);
+        assert!(option::contains(&account_resource.signer_capability_offer.for, &addr) || option::contains(&account_resource.rotation_capability_offer.for, &addr), error::not_found(EINVALID_ACCEPT_SIGNER_CAPABILITY));
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:404]: Uhm, I thought the signer capability was necessary to create a signer. It seems like the rotation capability suffices too though.

Why do we need two different capabilities for the same action: creating a signer? @davidiw

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 23]
```diff
@@ -515,27 +579,30 @@ module aptos_framework::account {
         rotate_authentication_key_ed25519(&alice, test_signature, test_signature, pk, pk);
     }
 
+    #[test_only]
+    public entry fun create_account_from_public_key(pk_bytes: vector<u8>): signer {
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:583]: A better name: `create_account_from_ed25519_public_key` lest other people think it's okay to call this with different types of PKs.

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 22 above)
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:404]: Because of this, it seems like a recipient that has the capability to rotate an offerer is able not just to rotate the offerer's key but also do other things, like transact?

Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:404]: i think you're right, i accidentally added it here. updating now : )

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 24]
```diff
@@ -323,7 +330,7 @@ module aptos_framework::account {
         option::fill(&mut account_resource.rotation_capability_offer.for, recipient_address);
     }
 
-    // Accept rotation capability from `offerer_address` if there's an existing rotation capability offer to the account owner in the offerer's account
+    /// Accept rotation capability from `offerer_address` if there's an existing rotation capability offer to the account owner in the offerer's account
     public fun accept_rotation_capability_ed25519(account: &signer, offerer_address: address) : RotationCapability acquires Account {
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:334]: On line 340 in this function, you need to wrap the error inside something from `move-stdlib/sources/error.move`, as pointed out before by @movekevin, no?

Can you make sure this holds for all `assert!` calls in this file?

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 25]
```diff
@@ -342,6 +349,63 @@ module aptos_framework::account {
...
+            assert!(ed25519::signature_verify_strict_t(&signer_capability_sig, &pubkey, proof_challenge), error::invalid_argument(EINVALID_PROOF_OF_KNOWLEDGE));
+        } else if (account_scheme == MULTI_ED25519_SCHEME) {
+            let pubkey = multi_ed25519::new_unvalidated_public_key_from_bytes(account_public_key_bytes);
+            let expected_auth_key = multi_ed25519::unvalidated_public_key_to_authentication_key(&pubkey);
+            assert!(account_resource.authentication_key == expected_auth_key, error::invalid_argument(EWRONG_CURRENT_PUBLIC_KEY));
+
+            let signer_capability_sig = multi_ed25519::new_signature_from_bytes(signer_capability_sig_bytes);
+            assert!(multi_ed25519::signature_verify_strict_t(&signer_capability_sig, &pubkey, proof_challenge), error::invalid_argument(EINVALID_PROOF_OF_KNOWLEDGE));
+        } else {
+            abort EINVALID_SCHEME
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:386]: Shouldn't this be wrapped in `error::invalid_argument`?

**Review by alinush — APPROVED** [review]
